crates/gas/src/call_graph/builder.rs

```diff
@@ -0,0 +1,69 @@
+use std::cell::{Ref, RefMut};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::rc::Rc;
+
+use crate::FuncIndex;
+
+use super::{CallGraph, Node, NodeRef, Value};
+
+/// This struct is used while building the `Call-Graph` on-the-fly as part of validation process of a Wasm program.
+/// Once the building is done, calling `build` will output a `CallGraph` struct.
+pub struct CallGraphBuilder<T = FuncIndex> {
+    nodes: HashMap<T, NodeRef<T>>,
+}
+
+impl<T> CallGraphBuilder<T>
+where
+    T: Value,
+{
+    pub fn new() -> Self {
+        Self {
+            nodes: HashMap::new(),
+        }
+    }
+
+    pub fn add_target(&mut self, value: T) {
+        let _ = self.get_or_create_mut(value);
+    }
+
+    pub fn add_call(&mut self, source: T, dest: T) {
+        debug_assert!(source != dest);
+
+        self.add_outgoing(source, dest);
+        self.add_incoming(source, dest);
+    }
+
+    pub fn build(self) -> CallGraph<T> {
```

Comment:
Since building is done on-the-fly, prior to this call, perhaps a different name is more appropriate.

Reply:
it's a convention in Rust to use `Builder` - i.e we add changes and once we're done we call `build`